src/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -750,6 +768,15 @@ private static void CheckBaseAddress(Uri baseAddress, string parameterName)
 
         private HttpRequestMessage CreateRequestMessage(HttpMethod method, Uri uri) =>
             new HttpRequestMessage(method, uri) { Version = _defaultRequestVersion };
+
+        private Exception CreateTimeoutException(Exception innerException)
+        {
+            return new HttpRequestException(
+                SR.net_http_timeout,
+                new TimeoutException(SR.net_http_timeout, innerException),
```

Comment:
Wondering whether adding 3 layers of exceptions is necessary here. Could we possibly eliminate either the TimeoutException or the innermost TCO?

Reply:
Well, introducing the TimeoutException (or at least an exception that makes it possible to distinguish between timeout and cancellation) is the whole point of this PR...

Reply:
I guess the question then is whether we need to surface the underlying TCO in that case.

Reply:
Yes, we could probably remove it. We'd loose the original stack trace, though... Not sure how much of a problem this is.

Reply:
@davidsh @stephentoub opinions?

Reply:
> I predict this is going to break a ton of code. Let's be really, really sure the benefit is worth the cost before going ahead with this. And explore alternatives that don't involve changing the exception type from an OperationCanceledException to a TimeoutException, if the goal is really to alert the caller that the reason for the cancellation is a timeout. We could change TimeoutException to derive from OperationCanceledException, which it probably should have in the first place... that is also likely to cause code to break, though I wonder if less.

Another option, to avoid a breaking change, is to introduce a `HttpTimeoutException` that derives from `OperationCanceledException` (as suggested by @scalablecory in https://github.com/dotnet/corefx/issues/20296#issuecomment-525596530). That seems less risky than changing the inheritance hierarchy of `TimeoutException`

Reply:
> Another option, to avoid a breaking change, is to introduce a HttpTimeoutException that derives from OperationCanceledException (as suggested by @scalablecory Cory Nelson FTE in #20296 (comment)). That seems less risky than changing the inheritance hierarchy of TimeoutException

It would have to inherit from `TaskCanceledException` rather than just `OperationCanceledException` to avoid breaking.

Reply:
I have worked on quite a few codebases that catch `TaskCanceledException`. This happens because people tend to add retry logic in reaction to error logs they see in prod rather than following best practices (in most cases people don't even realize that a subtype relationship exists).

Reply:
Ideally, it would be better to inherit from `OperationCanceledException`, but @eiriktsarpalis has a point: many people specifically catch `TaskCanceledException`. So I guess it's a choice between making the right design choice and limiting breaking changes...